### ISSUE TITLE
Filer sync seems stuck with "unexpected3 offset 0" error messages

### ISSUE THREAD

**Issue by douardda** [seaweedfs/seaweedfs#4112]
**Describe the bug**

On a seaweedfs cluster with 3 machines, each running master, (leveldb) filer and volume (no s3 nor fuse mount).

The cluster is used as an append only object storage, so no file object is ever modified or updated. All the filer services are used by the business software to write new objects (for load balancing and HA).

After some maintenance (os update and reboot) and the addition of a 4th machine in the cluster, the filer replication process seems stuck, with the following entrie in all 3 filer logs:

```
# logs from node 10.0.0.11, only the filers on 10.0.0.13 and 10.0.0.14 (new node in the cluster) are up and running at this point
# filer on  10.0.0.14 does not start syncing either, but shows no error in its logs
Jan 05 14:41:03 shsrvs1 seaweedfs-filer[226688]: I0105 14:41:03.092942 meta_aggregator.go:97 loopSubscribeToOneFiler read 10.0.0.13:8888 start from 2023-01-05 14:36:11.179052075 +0000 UTC 1672929371179052075
Jan 05 14:41:03 shsrvs1 seaweedfs-filer[226688]: I0105 14:41:03.094003 meta_aggregator.go:259 readOffset 10.0.0.13:8888 : 1671470021606448161
Jan 05 14:41:03 shsrvs1 seaweedfs-filer[226688]: I0105 14:41:03.094047 meta_aggregator.go:153 follow peer: 10.0.0.13:8888, last 2022-12-19 17:13:41.606448161 +0000 UTC (1671470021606448161)
Jan 05 14:41:03 shsrvs1 seaweedfs-filer[226688]: I0105 14:41:03.094084 meta_aggregator.go:194 subscribing remote 10.0.0.13:8888 meta change: 2022-12-19 17:13:41.606448161 +0000 UTC, clientId:548072232
Jan 05 14:41:03 shsrvs1 seaweedfs-filer[226688]: I0105 14:41:03.105254 meta_aggregator.go:108 subscribing remote 10.0.0.13:8888 meta change: rpc error: code = Unknown desc = reading from persisted logs: reading /topics/.system/log/2022-12-19/17-50.10fa7b93: unexpected3 offset 0 in 3126,823251392379c89b [208463,337584)
Jan 05 14:41:03 shsrvs1 seaweedfs-filer[226688]: I0105 14:41:03.205514 filer_grpc_server_sub_meta.go:279 + listener filer:10.0.0.13:8888@10.0.0.13:43772
Jan 05 14:41:03 shsrvs1 seaweedfs-filer[226688]: I0105 14:41:03.205549 filer_grpc_server_sub_meta.go:106  + filer:10.0.0.13:8888@10.0.0.13:43772 local subscribe / from 2022-11-23 15:59:04.473756458 +0000 UTC clientId:-1487099903
Jan 05 14:41:03 shsrvs1 seaweedfs-filer[226688]: I0105 14:41:03.205609 filer_grpc_server_sub_meta.go:119 read on disk filer:10.0.0.13:8888@10.0.0.13:43772 local subscribe / from 2022-11-23 15:59:04.473756458 +0000 UTC
Jan 05 14:41:03 shsrvs1 seaweedfs-filer[226688]: I0105 14:41:03.214777 filer_grpc_server_sub_meta.go:122 read on disk filer:10.0.0.13:8888@10.0.0.13:43772 local subscribe / from 2022-11-23 15:59:04.473756458 +0000 UTC: reading /topics/.system/log/2022-11-23/16-01.4fc8b5ed: unexpected3 offset 0 in 1368,450930d259ac7328 [1465,1392548)
Jan 05 14:41:03 shsrvs1 seaweedfs-filer[226688]: I0105 14:41:03.214817 filer_grpc_server_sub_meta.go:101  - filer:10.0.0.13:8888@10.0.0.13:43772 local subscribe / clientId:-1487099903
Jan 05 14:41:03 shsrvs1 seaweedfs-filer[226688]: I0105 14:41:03.214833 filer_grpc_server_sub_meta.go:294 - listener filer:10.0.0.13:8888@10.0.0.13:43772
```

The logs snippet is repeated once every 2 seconds as long as two of the 3 filer nodes are running.

**System Setup**

- weed 3.38: version 30GB 3.38 48f2edc06503d9a915fe8eb4cecbfc83f1467ce9 linux amd64
- OS: ubuntu jammy
- weed filer cmd line: `/usr/local/bin/weed filer -options=/etc/seaweedfs/filer.conf`
- filer.conf file:
```
# seaweedfs filer options
logdir=/swh/seaweedfs/log/filer
master=10.0.0.11:9333,10.0.0.12:9333,10.0.0.13:9333
ip=10.0.0.11 
ip.bind=0.0.0.0 
port=8888
```

- filer.toml:
```
# A sample TOML config file for SeaweedFS filer store
# Used with "weed filer" or "weed server -filer"
# Put this file to one of the location, with descending priority
#    ./filer.toml
#    $HOME/.seaweedfs/filer.toml
#    /etc/seaweedfs/filer.toml

####################################################
# Customizable filer server options
####################################################
[filer.options]
# with http DELETE, by default the filer would check whether a folder is empty.
# recursive_delete will delete all sub folders and files, similar to "rm -Rf"
recursive_delete = false

####################################################
# The following are filer store options
####################################################

[leveldb2]
# local on disk, mostly for simple single-machine setup, fairly scalable
# faster than previous leveldb, recommended.
enabled = true
dir = "/swh/seaweedfs/filerldb2"                    # directory to store level db files
```

**Expected behavior**

The filer synchronisation should resume properly and sync the leveldb indexes.

**Additional context**

Few related remarks/questions:
- is it possible to manually check the "synchronicity" status of 2 filers (list entries missing from one filer to the other)?
- I started with a multi active-active leveldb based filer solution for the sake of deployment simplicity, but I am thinking about migrating to a redis based solution, but I need to make sure current leveldb bases are synced before thinking of running  fs.meta.save/fs.meta.load . 
- Or can I save the 3 filer meta indexes in 3  files (they should have most of their content replicated), then run fs.meta.load from all these files in a single (eg. redis based) filer?
 

Thanks a lot!

**Comment by chrislusf** [comment]
> * Or can I save the 3 filer meta indexes in 3  files (they should have most of their content replicated), then run fs.meta.load from all these files in a single (eg. redis based) filer?

For your setup, yes.

**Comment by douardda** [comment]
Thanks a lot @chrislusf !

**Comment by douardda** [comment]
I confirm the sync between filers has resumed properly with this fix. Thanks

**Comment by chrislusf** [comment]
thanks for the confirmation!

**Comment by meotimdihia** [comment]
@douardda how did you fix this bug: "unexpected3 offset 0"?

![image](https://github.com/seaweedfs/seaweedfs/assets/300961/770b1ca2-d5be-4414-994b-65d77f68ea78)

**Comment by douardda** [comment]
> Edit: Fixed this problem by upgrading Seaweedfs to 3.50

was about to answer "I've upgraded weed" :-)
